### ISSUE TITLE
Discussion: Runtime refactoring

### ISSUE THREAD

**Issue by topherbullock** [concourse/concourse#2926]
(This may turn into an RFC for a more refined & well defined runtime interface.)

Recently there have been a few examples of features which have been difficult to implement because of the structure of the `worker` package of ATC.

- #2577
- #2910
- ...

We should look at the architecture of the code, and the interfaces available to create containers/volumes, find containers/volumes, select a worker for containers/volumes, etc.

**Comment by topherbullock** [comment]
Some notes I have on the `worker` package:

The [`worker.Worker` interface](https://github.com/concourse/atc/blob/df20fa592253abdd46c025a076fdb6be7da64e71/worker/worker.go#L47-L67)'s responsibility is a little unclear.

Two implementations are [`worker.gardenWorker`](https://github.com/concourse/atc/blob/df20fa592253abdd46c025a076fdb6be7da64e71/worker/worker.go#L69) and [`worker.pool`](https://github.com/concourse/atc/blob/df20fa592253abdd46c025a076fdb6be7da64e71/worker/pool.go)

Given these two implementations and the fact that many of the member methods *don't even work* for both ( [eg](https://github.com/concourse/atc/blob/df20fa592253abdd46c025a076fdb6be7da64e71/worker/pool.go#L194), [eg](https://github.com/concourse/atc/blob/df20fa592253abdd46c025a076fdb6be7da64e71/worker/worker.go#L265) ) implementations; it stands to reason that there's really two responsibilities being expressed: *Worker Selection* and *Operations on a Single Worker*
..
We also assert that `worker.pool` implements  `worker.Client`, which seems bonkers.

**Comment by marco-m** [comment]
Hello, first of all, thanks for all the work around scheduling :-)

I would like to give some input as the originator of #2577 to help prioritize this refactoring work.

Looking at `least-build-containers` (#2577), it will help us a lot. On the other hand, until #2928 (Add ATC/worker flags to limit max build containers for workers) (blocked by this refactoring) is done, we will still need to keep the workers alive by "controlling" the load by (ab)using the Concourse resource pool resource (I wrote a post at https://discuss.concourse-ci.org/t/ann-concourse-pool-boy-detect-and-release-stale-pool-resource-locks/765 describing the unavoidable flakiness of this approach).

Thanks again!

**Comment by topherbullock** [comment]
@marco-m Yeah, we're hoping this refactor will make the work for #2928 ( and all future work to improve the runtime ) much easier to complete; hopefully this area of the code will be easier to contribute to via PRs as well. 

Currently the code is a little daunting and hard to trace through. Especially due to things like this :

https://github.com/concourse/concourse/blob/6e4e81915b53fdd67b16fcdf76055d01811660eb/atc/worker/container_provider.go#L376-L384

Its not super obvious what that constructor is doing with container provider, and its passing the caller into the constructor.. SUPER WEIRD.. and what is that zero for !? (its an optimization for #2577 to pass down the number of build containers when constructing worker objects).

We ideally want to split up, or trim down the interfaces at play in implementing these features, so that we can easily extend them without impedance.

**Comment by pivotal-jwinters** [comment]
Another thing that we should consider is decoupling lifecycle of volumes from the lifecycle of containers. The [FindOrCreateContainer](https://github.com/concourse/concourse/blob/5c65ecfd18a8d248f87038561ba3bb1233ef6f3f/atc/worker/container_provider.go#L85) method does A LOT of stuff, including creating some volumes that it might need. 

There are [downstream operations](https://github.com/concourse/concourse/blob/master/atc/exec/task_step.go#L428) which try and figure out what volumes the [FindOrCreateContainer](https://github.com/concourse/concourse/blob/5c65ecfd18a8d248f87038561ba3bb1233ef6f3f/atc/worker/container_provider.go#L85) method might have created, so that it can associate the volume with another resource (e.g. task cache, or resource cache). This happens so the volume doesn't get gc'ed after the container goes away.

If we created all the volumes ahead of time we could decouple the task execution from the volume creation and we wouldn't have to figure out what volumes got created after the fact.

**Comment by topherbullock** [comment]
some more logs for the fire here; there are 2 functions that have high cyclomatic complexity in `container_provider.go`, according to `gocyclo`
```
23 worker (*containerProvider).createGardenContainer ./container_provider.go:314:1
18 worker (*containerProvider).FindOrCreateContainer ./container_provider.go:82:1
```

And both the test setup and constructor have a lot of dependancies. 

Here's some good spots to look at for a refactor at the moment:
- [ ] splitting that method up into smaller parts
- [ ] passing a set of volumes (or maybe, an image volume and a set of bind mounts?) to `FindOrCreateContainer` and dealing with volume creation at the call site.
- [ ] reconsider what purpose `ContainerProvider` serves and what it relies on to fill that purpose; is  "provides containers" an unclear purpose (esp given it will likely no longer depend on volumeClient, or imageFactory)?

**Comment by ddadlani** [comment]
Small note: This issue can still be used for discussion. To track progress, the chunks of work that arise from this discussion will be captured in other issues (such as #3043) which will be linked back here.

**Comment by ddadlani** [comment]
@marco-m No problem! :)

The next issue we are trying to tackle is the creation of volumes. Currently, volumes are created in [containerProvider.createGardenContainer](https://github.com/concourse/concourse/blob/8f9d96b25a3f7813a0c936a029d41d43e2afc454/atc/worker/container_provider.go#L304). The reason for this is that you need a container handle to create volumes in the database, which we can only obtain from the `creatingContainer`. 

However, this makes it hard to pull out volume creation to an earlier stage (as per https://github.com/concourse/concourse/issues/2926#issuecomment-449149707). It also means that the `containerProvider` requires access to a `volumeClient`, which is odd because the `atc/worker` also has a `volumeClient` and `containerProvider` methods are called from `atc/worker` methods.

It almost seems like a big reason for this confusion is that `worker` and `containerProvider` are sharing responsibilities that are not easy to split. For this reason, we are going to try moving all the container/volume creation logic to the `worker` and then see what can be split out into the `containerProvider`, potentially making the `containerProvider` a collection of helper functions for the worker.

**Comment by ddadlani** [comment]
Another interesting idea from @topherbullock was that we could maybe have a `containerProvider` "helper" equivalent for `db/worker` as well, which handles the database operations required to create containers. Hopefully this would make things easier at a later stage.

**Comment by ddadlani** [comment]
Update in child issue: https://github.com/concourse/concourse/issues/3052#issuecomment-456184038

**Comment by marco-m** [comment]
Hello @ddadlani, do you have a high-level view of what will come next ? I would like to understand what is in the way of #2928. Thanks!

**Comment by ddadlani** [comment]
Hi @marco-m, we hadn't planned to implement #2928 directly after this issue, it hasn't even been prioritized yet. Could you maybe give us more context on #2928 in the other issue? I posted a question there as well.

To answer your question, we are currently refactoring `atc/worker/worker.go` and `container_provider.go` to simplify container and volume creation, and are considering ways of simplifying the volume lifecycle as part of #3079.

**Comment by marco-m** [comment]
@hello @ddadlani, thanks for the status update, greatly appreciated :-) I will answer in #2928.

**Comment by vito** [comment]
This has led to other stuff and is otherwise pretty open-ended so we're gonna call this good for now. 👍
